packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,159 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
-        continue;
-      }
-      // Next, try direct dependencies of the resolving application.
-      final Iterable<Plugin> directDependencies = candidates.where((Plugin plugin) {
-        return plugin.isDirectDependency;
-      });
-      if (directDependencies.isNotEmpty) {
-        if (directDependencies.length > 1) {
-          globals.printError(
-              'Plugin ${entry.key}:$platformKey has conflicting direct dependency implementations:\n'
-                  '${directDependencies.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                  'To fix this issue, remove all but one of these dependencies from pubspec.yaml.\n'
-          );
-          hasResolutionError = true;
-        } else {
-          pluginResolution.add(directDependencies.first);
+    // Now resolve all the possible resolutions to a single option for each
+    // plugin, or throw if that's not possible.
+    for (final MapEntry<String, List<Plugin>> implCandidatesEntry in pluginImplCandidates.entries) {
+      try {
+        final Plugin? resolution = _resolveImplementationOfPlugin(
+          platformKey,
+          implCandidatesEntry,
+          defaultPackageName: defaultImplementations[implCandidatesEntry.key],
+        );
+        if (resolution != null) {
+          pluginResolution[implCandidatesEntry.key] = resolution;
         }
-        continue;
-      }
-      // Next, defer to the default implementation if there is one.
-      final String? defaultPackageName = defaultImplementations[entry.key];
-      if (defaultPackageName != null) {
-        final int defaultIndex = candidates
-            .indexWhere((Plugin plugin) => plugin.name == defaultPackageName);
-        if (defaultIndex != -1) {
-          pluginResolution.add(candidates[defaultIndex]);
-          continue;
+      } on ToolExit catch (e) {
+        if (e.message != null) {
+          globals.printError(e.message!);
         }
-      }
-      // Otherwise, require an explicit choice.
-      if (candidates.length > 1) {
-        globals.printError(
-            'Plugin ${entry.key}:$platformKey has multiple possible implementations:\n'
-                '${candidates.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
-        );
         hasResolutionError = true;
-        continue;
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+/// Get the [resolutionPluginName] as the first choice for the plugin
+/// implementation and the [defaultImplementation] as fallback.
```

Comment:
I'm not entirely sure from this comment what "the plugin implementation" means here. Is it "the implementation for the app-facing plugin [plugin]"?

Reply:
More generally I don't think I understand how a method that does what this says can exist with these arguments. It's impossible to know what the first choice implementation of a given plugin is when looking at that plugin in isolation; that's why all the resolution logic exists in the first place.

Reply:
Done in 12a38f9921d0390979945fcfe7e04f03842f2246.